File: modules/benchmarks-ts/src/circles.ts
//! STDB module used for benchmarks based on "realistic" workloads we are focusing in improving.

import { type Load, newLoad, blackBox } from './load';
import { spacetimedb, type Entity, type Circle, type Food } from './schema';
import { Timestamp } from 'spacetimedb';
import { t } from 'spacetimedb/server';

function newEntity(id: number, x: number, y: number, mass: number): Entity {
  return {
    id,
    position: { x, y },
    mass,
  };
}

function newCircle(
  entity_id: number,
  player_id: number,
  x: number,
  y: number,
  magnitude: number,
  last_split_time: Timestamp
): Circle {
  return {
    entity_id,
    player_id,
    direction: { x, y },
    magnitude,
    last_split_time,
  };
}

function newFood(entity_id: number): Food {
  return {
    entity_id,
  };
}

function massToRadius(mass: number): number {
  return Math.sqrt(mass);
}

function isOverlapping(entity1: Entity, entity2: Entity): boolean {
  const entity1Radius = massToRadius(entity1.mass);
  const entity2Radius = massToRadius(entity2.mass);
  const distance = Math.sqrt(
    (entity1.position.x - entity2.position.x) ** 2 +
    (entity1.position.y - entity2.position.y) ** 2
  );
  return distance < Math.max(entity1Radius, entity2Radius);
}

// ---------- insert bulk ----------

const insertBulkEntity = spacetimedb.reducer('insert_bulk_entity', { count: t.u32() }, (ctx, { count }) => {
  for (let id = 0; id < count; id++) {
    ctx.db.entity.insert(newEntity(0, id, id + 5, id * 5));
  }
  console.info(`INSERT ENTITY: ${count}`);
});

const insertBulkCircle = spacetimedb.reducer('insert_bulk_circle', { count: t.u32() }, (ctx, { count }) => {
  for (let id = 0; id < count; id++) {
    ctx.db.circle.insert(newCircle(id, id, id, id + 5, id * 5, ctx.timestamp));
  }
  console.info(`INSERT CIRCLE: ${count}`);
});

const insertBulkFood = spacetimedb.reducer('insert_bulk_food', { count: t.u32() }, (ctx, { count }) => {
  for (let id = 1; id <= count; id++) {
    ctx.db.food.insert(newFood(id));
  }
  console.info(`INSERT FOOD: ${count}`);
});

// Simulate
// ```
// SELECT * FROM Circle, Entity, Food
// ```
const crossJoinAll = spacetimedb.reducer('cross_join_all', { expected: t.u32() }, (ctx, { expected }) => {
  let count: number = 0;

  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  for (const circle of ctx.db.circle.iter()) {
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    for (const entity of ctx.db.entity.iter()) {
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
      for (const food of ctx.db.food.iter()) {
        count += 1;
      }
    }
  }

  console.info(`CROSS JOIN ALL: ${expected}, processed: ${count}`);
});

// Simulate
// ```
// SELECT * FROM Circle JOIN ENTITY USING(entity_id), Food JOIN ENTITY USING(entity_id)
// ```
const crossJoinCircleFood = spacetimedb.reducer(
  'cross_join_circle_food',
  { expected: t.u32() },
  (ctx, { expected }) => {
    let count: number = 0;

    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    for (const circle of ctx.db.circle.iter()) {
      const entityId = ctx.db.entity.id;
      const circleEntity = entityId.find(circle.entity_id);
      if (circleEntity == null) {
        continue;
      }

      for (const food of ctx.db.food.iter()) {
        count += 1;

        const foodEntity = entityId.find(food.entity_id);
        if (foodEntity == null) {
          let string = JSON.stringify(circleEntity);
          throw new Error(`Entity not found: ${food.entity_id}`);
        }

        blackBox(isOverlapping(circleEntity, foodEntity));
      }
    }

    console.info(`CROSS JOIN CIRCLE FOOD: ${expected}, processed: ${count}`);
  }
);

spacetimedb.reducer(
  'init_game_circles',
  { initial_load: t.u32() },
  (ctx, { initial_load }) => {
    const load = newLoad(initial_load);

    insertBulkFood(ctx, { count: load.initialLoad });
    insertBulkEntity(ctx, { count: load.initialLoad });
    insertBulkCircle(ctx, { count: load.smallTable });
  }
);

spacetimedb.reducer(
  'run_game_circles',
  { initial_load: t.u32() },
  (ctx, { initial_load }) => {
    const load = newLoad(initial_load);

    crossJoinCircleFood(ctx, { expected: initial_load * load.smallTable });
    crossJoinAll(ctx, {
      expected: initial_load * initial_load * load.smallTable,
    });
  }
);

File: modules/benchmarks-cpp/src/circles.cpp
//! STDB module used for benchmarks based on "realistic" workloads we are focusing in improving.
//! Circles benchmark - Game-like entities with spatial queries

#include "common.h"
#include <cmath>
#include <algorithm>

// =============================================================================
// CIRCLES BENCHMARK - DATA STRUCTURES
// =============================================================================

// Vector2 - 2D coordinate structure
struct Vector2 {
    float x;
    float y;
};
SPACETIMEDB_STRUCT(Vector2, x, y)

// Entity table - represents game objects with position and mass
struct Entity {
    uint32_t id;
    Vector2 position;
    uint32_t mass;
};
SPACETIMEDB_STRUCT(Entity, id, position, mass)
SPACETIMEDB_TABLE(Entity, entity, Public)
FIELD_PrimaryKeyAutoInc(entity, id)

// Circle table - represents player-controlled entities
struct Circle {
    uint32_t entity_id;
    uint32_t player_id;
    Vector2 direction;
    float magnitude;
    Timestamp last_split_time;
};
SPACETIMEDB_STRUCT(Circle, entity_id, player_id, direction, magnitude, last_split_time)
SPACETIMEDB_TABLE(Circle, circle, Public)
FIELD_PrimaryKey(circle, entity_id)
FIELD_Index(circle, player_id)

// Food table - represents consumable game objects
struct Food {
    uint32_t entity_id;
};
SPACETIMEDB_STRUCT(Food, entity_id)
SPACETIMEDB_TABLE(Food, food, Public)
FIELD_PrimaryKey(food, entity_id)

// =============================================================================
// CIRCLES BENCHMARK - HELPER FUNCTIONS
// =============================================================================

// Convert mass to radius for collision detection
inline float mass_to_radius(uint32_t mass) {
    return std::sqrt(static_cast<float>(mass));
}

// Check if two entities are overlapping based on their positions and masses
inline bool is_overlapping(const Entity& entity1, const Entity& entity2) {
    float entity1_radius = mass_to_radius(entity1.mass);
    float entity2_radius = mass_to_radius(entity2.mass);
    float dx = entity1.position.x - entity2.position.x;
    float dy = entity1.position.y - entity2.position.y;
    float distance = std::sqrt(dx * dx + dy * dy);
    return distance < std::max(entity1_radius, entity2_radius);
}

// =============================================================================
// CIRCLES BENCHMARK - BULK INSERT OPERATIONS
// =============================================================================

// Bulk insert entities with auto-incremented IDs
SPACETIMEDB_REDUCER(insert_bulk_entity, ReducerContext& ctx, uint32_t count) {
    for (uint32_t id = 0; id < count; ++id) {
        Entity new_entity = {
            0, // Auto-incremented by database
            {static_cast<float>(id), static_cast<float>(id + 5)}, // position
            id * 5 // mass
        };
        ctx.db[entity].insert(new_entity);
    }
    LOG_INFO("INSERT ENTITY: " + std::to_string(count));
    return Ok();
}

// Bulk insert circles with specified entity and player IDs
SPACETIMEDB_REDUCER(insert_bulk_circle, ReducerContext& ctx, uint32_t count) {
    for (uint32_t id = 0; id < count; ++id) {
        Circle new_circle = {
            id, // entity_id
            id, // player_id
            {static_cast<float>(id), static_cast<float>(id + 5)}, // direction
            static_cast<float>(id * 5), // magnitude
            ctx.timestamp // last_split_time
        };
        ctx.db[circle].insert(new_circle);
    }
    LOG_INFO("INSERT CIRCLE: " + std::to_string(count));
    return Ok();
}

// Bulk insert food entities
SPACETIMEDB_REDUCER(insert_bulk_food, ReducerContext& ctx, uint32_t count) {
    for (uint32_t id = 1; id <= count; ++id) {
        Food new_food = {id};
        ctx.db[food].insert(new_food);
    }
    LOG_INFO("INSERT FOOD: " + std::to_string(count));
    return Ok();
}

// =============================================================================
// CIRCLES BENCHMARK - CROSS JOIN OPERATIONS
// =============================================================================

// Simulate: SELECT * FROM Circle, Entity, Food
// This creates a Cartesian product of all three tables
SPACETIMEDB_REDUCER(cross_join_all, ReducerContext& ctx, uint32_t expected) {
    uint32_t count = 0;
    for (const auto& _circle : ctx.db[circle]) {
        for (const auto& _entity : ctx.db[entity]) {
            for (const auto& _food : ctx.db[food]) {
                ++count;
            }
        }
    }
    LOG_INFO("CROSS JOIN ALL: " + std::to_string(expected) + ", processed: " + std::to_string(count));
    return Ok();
}

// Simulate: SELECT * FROM Circle JOIN Entity USING(entity_id), Food JOIN Entity USING(entity_id)
// This joins circles with their entities, then cross-joins with food entities to check overlaps
SPACETIMEDB_REDUCER(cross_join_circle_food, ReducerContext& ctx, uint32_t expected) {
    uint32_t count = 0;
    for (const auto& circle_elem : ctx.db[circle]) {
        // Find the entity for this circle
        auto circle_entity_opt = ctx.db[entity_id].find(circle_elem.entity_id);
        if (!circle_entity_opt) {
            continue; // Skip if entity not found
        }
        const auto& circle_entity = *circle_entity_opt;
        
        // Cross join with all food entities
        for (const auto& food_elem : ctx.db[food]) {
            ++count;
            // Find the entity for this food
            auto food_entity_opt = ctx.db[entity_id].find(food_elem.entity_id);
            if (!food_entity_opt) {
                return Err("Entity not found: " + std::to_string(food_elem.entity_id));
            }
            const auto& food_entity = *food_entity_opt;
            
            // Check overlap and use black_box to prevent optimization
            black_box(is_overlapping(circle_entity, food_entity));
        }
    }
    LOG_INFO("CROSS JOIN CIRCLE FOOD: " + std::to_string(expected) + ", processed: " + std::to_string(count));
    return Ok();
}

// =============================================================================
// CIRCLES BENCHMARK - GAME SIMULATION ENTRY POINTS
// =============================================================================

// Initialize the circles game simulation with test data
SPACETIMEDB_REDUCER(init_game_circles, ReducerContext& ctx, uint32_t initial_load) {
    Load load(initial_load);
    
    // Set up the game world with food, entities, and circles
    auto bulk_food_res = insert_bulk_food(ctx, load.initial_load);
    if (bulk_food_res.is_err()) {
        return bulk_food_res;
    }
    auto bulk_entity_res = insert_bulk_entity(ctx, load.initial_load);
    if (bulk_entity_res.is_err()) {
        return bulk_entity_res;
    }
    auto bulk_circle_res = insert_bulk_circle(ctx, load.small_table);
    if (bulk_circle_res.is_err()) {
        return bulk_circle_res;
    }
    return Ok();
}

// Run the circles game simulation benchmark
SPACETIMEDB_REDUCER(run_game_circles, ReducerContext& ctx, uint32_t initial_load) {
    Load load(initial_load);
    
    // Perform the main benchmark operations
    auto cross_join_circle_food_res = cross_join_circle_food(ctx, initial_load * load.small_table);
    if (cross_join_circle_food_res.is_err()) {
        return cross_join_circle_food_res;
    }
    auto cross_join_all_res = cross_join_all(ctx, initial_load * initial_load * load.small_table);
    if (cross_join_all_res.is_err()) {
        return cross_join_all_res;
    }
    return Ok();
}

